File: Python語法解釋/oop-polymorphism.py
from abc import ABC, abstractmethod

# ==========================================
# 1. 抽象類別 Hero (展現：抽象 & 封裝)
# ==========================================
class Hero(ABC):
    def __init__(self, name: str, hp: int, attack_power: int):
        self._name = name            # 封裝：使用 protected 底線開頭
        self._hp = hp
        self._max_hp = hp
        self._attack_power = attack_power

    def get_name(self) -> str:
        return self._name

    def is_alive(self) -> bool:
        return self._hp > 0

    def get_status(self) -> str:
        return f"[{self._name}] HP: {self._hp}/{self._max_hp}"

    def take_damage(self, amount: int) -> None:
        actual_damage = max(0, amount)
        self._hp = max(0, self._hp - actual_damage)
        print(f"  💥 {self._name} 受到了 {actual_damage} 點傷害！ (剩餘 HP: {self._hp}/{self._max_hp})")

    @abstractmethod
    def attack(self, target: 'Hero') -> None:
        """抽象方法：強制所有子類別必須實作攻擊邏輯"""
        pass


# ==========================================
# 2. 繼承類別 Warrior (展現：繼承 & 多型)
# ==========================================
class Warrior(Hero):
    def __init__(self, name: str, hp: int, attack_power: int, armor: int):
        super().__init__(name, hp, attack_power) # 呼叫父類別建構子
        self._armor = armor

    def take_damage(self, amount: int) -> None:
        # 覆寫受傷邏輯：先用護甲抵扣
        blocked = min(amount, self._armor)
        reduced_damage = max(0, amount - self._armor)
        print(f"  🛡️ {self._name} 的護甲擋下了 {blocked} 點傷害！")
        super().take_damage(reduced_damage)

    def attack(self, target: Hero) -> None:
        print(f"🗡️ [戰士] {self._name} 對 {target.get_name()} 發動砍擊！(基礎攻擊力: {self._attack_power})")
        target.take_damage(self._attack_power)


# ==========================================
# 3. 繼承類別 Mage (展現：繼承 & 多型)
# ==========================================
class Mage(Hero):
    def __init__(self, name: str, hp: int, attack_power: int, mp: int):
        super().__init__(name, hp, attack_power)
        self._mp = mp
        self._max_mp = mp

    def attack(self, target: Hero) -> None:
        if self._mp >= 10:
            self._mp -= 10
            magic_damage = self._attack_power * 2
            print(f"🔥 [法師] {self._name} 消耗 10 MP 發動大火球術！(剩餘 MP: {self._mp}/{self._max_mp})")
            target.take_damage(magic_damage)
        else:
            self._mp += 5
            print(f"✨ [法師] {self._name} MP 不足，改用吟唱杖擊並恢復 5 MP！(當前 MP: {self._mp}/{self._max_mp})")
            target.take_damage(self._attack_power)

    def get_status(self) -> str:
        return f"[{self._name}] HP: {self._hp}/{self._max_hp} | MP: {self._mp}/{self._max_mp}"


# ==========================================
# 4. 戰鬥模擬測試主程式
# ==========================================
if __name__ == "__main__":
    print("============== ⚔️ 英雄 RPG 戰鬥模擬開始 ⚔️ ==============\n")

    # 建立英雄隊伍 (多型展示：列表內同時存放 Warrior 與 Mage)
    party = [
        Warrior(name="亞瑟", hp=120, attack_power=20, armor=5),
        Mage(name="梅林", hp=80, attack_power=15, mp=15)
    ]

    boss = Warrior(name="首領石像鬼", hp=250, attack_power=10, armor=8)

    print("【隊伍初始狀態】")
    for hero in party:
        print(f"  - {hero.get_status()}")
    print(f"  - Boss: {boss.get_status()}\n")

    # 進行 3 回合戰鬥
    for round_num in range(1, 4):
        print(f"------------------ 第 {round_num} 回合 ------------------")
        for hero in party:
            if hero.is_alive() and boss.is_alive():
                hero.attack(boss)
                print("")

        print("【本回合結束戰況】")
        for hero in party:
            print(f"  {hero.get_status()}")
        print(f"  {boss.get_status()}\n")

    print("==================== 戰鬥結束 ====================")

File: Python語法解釋/oop-polymorphism.ts
// oop-polymorphism.ts
// 練習目標：抽象（Abstraction）、封裝（Encapsulation）、繼承（Inheritance）、多型（Polymorphism）
// 執行方式：npx tsx oop-polymorphism.ts   或貼到 TypeScript Playground

// ========== 1. 抽象類別 Hero：抽象 + 封裝 ==========
abstract class Hero {
  protected name: string;        // protected：子類別可存取，外部不行
  protected hp: number;
  protected maxHp: number;
  protected attackPower: number;

  constructor(name: string, hp: number, attackPower: number) {
    this.name = name;
    this.hp = hp;
    this.maxHp = hp;
    this.attackPower = attackPower;
  }

  public getName(): string { return this.name; }
  public isAlive(): boolean { return this.hp > 0; }
  public getStatus(): string { return `[${this.name}] HP: ${this.hp}/${this.maxHp}`; }

  public takeDamage(amount: number): void {
    const actual = Math.max(0, amount);
    this.hp = Math.max(0, this.hp - actual);   // 封裝：HP 永遠不會低於 0
    console.log(`  💥 ${this.name} 受到 ${actual} 點傷害（剩餘 HP: ${this.hp}/${this.maxHp}）`);
  }

  // 抽象方法：只立契約，強制子類別實作
  abstract attack(target: Hero): void;
}

// ========== 2. Warrior：繼承 + 覆寫 ==========
class Warrior extends Hero {
  private armor: number;

  constructor(name: string, hp: number, attackPower: number, armor: number) {
    super(name, hp, attackPower);
    this.armor = armor;
  }

  override takeDamage(amount: number): void {
    const blocked = Math.min(amount, this.armor);
    console.log(`  🛡️ ${this.name} 的護甲擋下 ${blocked} 點傷害`);
    super.takeDamage(Math.max(0, amount - this.armor));
  }

  attack(target: Hero): void {
    console.log(`🗡️ [戰士] ${this.name} 對 ${target.getName()} 砍擊（攻擊力 ${this.attackPower}）`);
    target.takeDamage(this.attackPower);
  }
}

// ========== 3. Mage：繼承 + 覆寫 ==========
class Mage extends Hero {
  private mp: number;
  private maxMp: number;

  constructor(name: string, hp: number, attackPower: number, mp: number) {
    super(name, hp, attackPower);
    this.mp = mp;
    this.maxMp = mp;
  }

  attack(target: Hero): void {
    if (this.mp >= 10) {
      this.mp -= 10;
      console.log(`🔥 [法師] ${this.name} 消耗 10 MP 發動火球術（剩餘 MP: ${this.mp}/${this.maxMp}）`);
      target.takeDamage(this.attackPower * 2);
    } else {
      this.mp += 5;
      console.log(`✨ [法師] ${this.name} MP 不足，改用杖擊並回 5 MP（當前 MP: ${this.mp}/${this.maxMp}）`);
      target.takeDamage(this.attackPower);
    }
  }

  override getStatus(): string {
    return `[${this.name}] HP: ${this.hp}/${this.maxHp} | MP: ${this.mp}/${this.maxMp}`;
  }
}

// ========== 4. 多型示範 ==========
function startBattleDemo(): void {
  console.log('============ ⚔️ 英雄 RPG 戰鬥模擬開始 ⚔️ ============\n');

  // 關鍵：陣列型別是 Hero[]，卻裝著 Warrior 與 Mage
  const party: Hero[] = [
    new Warrior('亞瑟', 120, 20, 5),
    new Mage('梅林', 80, 15, 15),
  ];
  const boss = new Warrior('首領石像鬼', 250, 10, 8);

  for (let round = 1; round <= 3; round++) {
    console.log(`---------------- 第 ${round} 回合 ----------------`);
    for (const hero of party) {
      if (hero.isAlive() && boss.isAlive()) {
        hero.attack(boss);   // ← 同一行程式碼，兩種行為，這就是多型
        console.log('');
      }
    }
    party.forEach(h => console.log(`  ${h.getStatus()}`));
    console.log(`  ${boss.getStatus()}\n`);
  }
  console.log('==================== 戰鬥結束 ====================');
}

startBattleDemo();
